documentation/ha.markdown

```diff
@@ -0,0 +1,265 @@
+
+Manual Installation and Configuration
+----
+
+1. Provision two PuppetDB nodes. Designate one of these as your _primary_
+   PuppetDB, and the other as the _replica_. Be sure that each PuppetDB is using
+   its own PostgreSQL cluster. HA doesn't buy you anything if you have single
+   database. In this guide, we will give them the hostnames
+   `primary-puppetdb.mycorp.net` and `replica-puppetdb.mycorp.net`.
+
+2. Configure each PuppetDB to pull from the other. This guide will use
+   HOCON-style configuration. See the
+   [pe-configure][PE Configuration Documentation] for further details.
+
+  a. Configure primary-puppetdb to pull from replica-puppetdb by placing this in its
+     config file:
+
+         sync: {
+           remotes: [{server_url: "https://replica-puppetdb.mycorp.net:8081",
+                      interval: 5m}]
+         }
+
+  b. Configure replica-puppetdb to pull from primary-puppetdb:
+
+         sync: {
+           remotes: [{server_url: "https://primary-puppetdb.mycorp.net:8081",
+                      interval: 5m}]
+         }
+
+3. Configure the Puppet master to point to both of your PuppetDB instances. In
+   your puppetdb.conf, use this as your server configuration (in the 'main' section):
+
+        server_urls: ["https://primary-puppetdb.mycorp.net:8081", "https://replica-puppetdb.mycorp.net:8081"]
+        sticky_read_failover: true
+        command_broadcast: true
+
+   It is important that your primary PuppetDB appear first in this list, so it
+   will always be tried first when submitting new data from the Puppet master.
+
+4. Restart your PuppetDBs and your Puppet master.
+
+Deployment tips
+----
+
+- Each PuppetDB server should be configured to use a separate PostgreSQL
+  instance.
+
+- If you are using multiple compile masters, be sure that their clocks are
+  relatively in sync (using NTP, for example). PuppetDB relies on timestamps for
+  discerning which data is newer for any given node, so the clocks on your
+  compile masters need to be within `runinterval` of each other.
+
+FAQ
+-----
+
+Q: *Why not just use PostgreSQL streaming replication?*
+
+A: The principal goal of PuppetDB's HA solution is to ensure that Puppet runs
+never fail, even in the face of losing machines and intermittent network
+connections. These precludes any possibility of coordinating writes among the
+members of the cluster. That is, this is a system that cares first about
+availability and second about consistency.
+
+Relational database replication systems tend to emphasize consistency. Those
+that are availability-focused must concern themselves with conflict resolution.
+For PuppetDB in particular, conflict resolution at the database level would be
+very complex and potentially intractable. Conversely, at the application level
+we can view the entities in very coarse terms (e.g. a whole catalog vs. the many
+database rows required to store its resources). This allows us to resolve any
+conflicts very predictably.
```

Comment:
FWIW the thing that ticked me off here was the "very predictably" bit. Most of the time "very" doesn't add much and will increase frustration when our statements turn out to be wrong.